THREESCALE-6894: Added updates for the procedure to configure the integration setting in 3scale.

dfennessy opened the pull request:
This pull requested is based on [THREESCALE-6894](https://issues.redhat.com/browse/THREESCALE-6894)

Reviewers

@unleashed
@neal-timpe

dfennessy commented:
This PR will have more updates added once I've received feedback from the SMEs based on some questions in the comments in THREESCALE-6894

unleashed commented on modules/ossm-threescale-integrate-1x.adoc:
```diff
@@ -24,7 +24,11 @@ To configure the 3scale Istio Adapter, refer to {ProductName} custom resources f
 [NOTE]
 ====
-Pay particular attention to the `kind: handler` resource. You must update this with your 3scale credentials and the service ID of the API you want to manage.
+Pay particular attention to the `kind: handler` resource. You must update this with your 3scale credentials and the `service_id` of the API you want to manage. For example:
+
+* Create one handler per service by attaching service IDs to workloads by labeling the relevant Deployment.
+* Create multiple different `service_id` workloads under the same 3scale account that refer to a single handler. This stores the 3scale credentials in their normal fields except `service_id`.
+* Then you only need one handler per 3scale account, rather than one handler per service of each 3scale account.
```
I'd rewrite this as follows (also consider this change for the identical text in ossm-threescale-integrate.adoc):

```suggestion
Pay particular attention to the `kind: handler` resource. You must update this with your 3scale account credentials. You can optionally add a `service_id` to a handler, but this is kept for backwards compatibility only, since it would render the handler only useful for one service in your 3scale account, and enabling 3scale for other services will require for you to create more handlers which only differ in their `service_id`s.

You are instead encouraged to use a single handler per 3scale account by following the steps below:

1. Create a handler for your 3scale account and specify your account credentials. Please omit any service identifier.
2. Edit or patch the Deployment resource of any services belonging to your 3scale account as follows:
  a. Add the `"service-mesh.3scale.net/service-id"` label with a value corresponding to a valid `service_id`.
  b. Add the `"service-mesh.3scale.net/credentials"` label with its value being the _name of the handler resource_ from step 1.
3. Whenever you further add services, perform step 2 to link it to your 3scale account credentials and to its service identifier.
```

dfennessy replied:
Thanks, @unleashed 

How does this affect the Handler configuration example and the Rule configuration example that follows?

unleashed replied:
The service_id entry would disappear from the params in the handler and instead the user would label their workload as specified in the steps above, so workloads would need to have two labels: one to point to the handler so the adapter knows how to contact their 3scale administration portal and account, and the other label to specify the service identifier within their account that that particular workload represents.

dfennessy replied:
@unleashed Thanks for your reply. I've made some updates. Hopefully they are as you described or we are close to it :smile:

unleashed replied:
I think it is there already, just added a minor remark on some wording to avoid confusion: https://github.com/openshift/openshift-docs/pull/32951#discussion_r647637711

dfennessy commented on modules/ossm-mixer-policy.adoc:
```diff
@@ -19,14 +19,14 @@ In previous versions of {ProductName}, Mixer’s policy enforcement was enabled
 [source,terminal]
 ----
-$ oc get cm -n istio-system istio -o jsonpath='{.data.mesh}' | grep disablePolicyChecks
+$ oc get cm -n istio-<smcp-name> istio -o jsonpath='{.data.mesh}' | grep disablePolicyChecks
```
Thanks, Alex.

So what you're saying is that we can leave that command as it was?

unleashed replied:
Yes, but somewhere in the text it should explain that this is the namespace where the Service Mesh Control Plane is installed, which may or may not be `istio-system`, as that value can be changed by the person installing it.

unleashed replied:
@dfennessy this still looks a bit odd to me - is there some clarification as to what this should mean?

dfennessy replied:
For some reason I don't see my changes showing up.

unleashed commented on modules/ossm-threescale-integrate-1x.adoc:
```diff
@@ -21,15 +20,17 @@ You can use these examples to configure requests to your services using the 3sca
 [NOTE]
 ====
-Pay particular attention to the `kind: handler` resource. You must update this with your 3scale credentials and the service ID of the API you want to manage.
+Pay particular attention to the `kind: handler` resource. You must update this with your 3scale account credentials. You can optionally add a `service_id` to a handler, but this is kept for backwards compatibility only, since it would render the handler only useful for one service in your 3scale account. Enabling 3scale for other services will require for you to create more handlers which only di ...
```
I'd rewrite the last sentence to make it clear that it will only happen if the user goes the route of specifying the service_id in the handler: `If you add service_id to a handler, enabling 3scale for other services will require for you to create more handlers which only differ in their service_id's.`

Edit: note this also applies as well to the same text in modules/ossm-threescale-integrate.adoc.

dfennessy commented:
Thanks @unleashed 

Updates made.

unleashed commented on modules/ossm-mixer-policy-1x.adoc:
```diff
@@ -26,7 +27,7 @@ $ oc get cm -n istio-system istio -o jsonpath='{.data.mesh}' | grep disablePolic
 [source,terminal]
 ----
-$ oc edit cm -n istio-system istio
+$ oc edit cm -n istio-<smcp-name> istio
```
This instance is still remaining, while it is fixed in `module/ossm-mixer-policy.adoc`.

dfennessy replied:
Cheers! Done...

unleashed reviewed (approved):
LGTM

pehala commented on modules/ossm-threescale-integration-settings.adoc:
```diff
@@ -15,10 +15,10 @@ For 3scale SaaS customers, {ProductName} is enabled as part of the Early Access
+. Select the *Istio* option under _Deployment_.
 
-. Scroll to the bottom of the page and click *Update Service*.
+. Click *Update Product* to save your selection.
```
I believe there is an additional step required:

- Go into *Configuration*
- Click *Update configuration*

pehala commented on modules/ossm-threescale-manifests.adoc:
```diff
@@ -6,6 +6,12 @@
+====
+* `NAME` is an identifier you use to identify with the service you are managing with 3scale.
+* The `CREDENTIALS_NAME` reference is an identifier that corresponds to the `match` section in the rule configuration. This is automatically set to the `NAME` identifier if you are using the CLI tool.
+* Its value does not need to be anything specific: the label value should just match the contents of the rule. See link:https://github.com/3scale/3scale-istio-adapter/blob/v2.0.2/README.md#routing-service-traffic-through-the-adapter[Routing service traffic through the adapter] for more information.
```
I am not sure if we should use links to a static version of upstream README as it may change with the new patch releases.

unleashed replied:
This is unlikely to ever change as the adapter will see no more releases other than maintenance in already shipping products. The correct fix would probably be to link to the `v2.X` or the `v2.0.X` branches (which would guarantee this), but then given a big change not just the link but the surrounding text would also be outdated and need changes. Not something I'd consider important, but feel free to link instead to: `https://github.com/3scale/3scale-istio-adapter/blob/v2.X/README.md`

pehala commented on modules/ossm-threescale-templates.adoc:
```diff
@@ -6,14 +6,25 @@
 +
 ----
-$ 3scale-gen-config --url="https://<organization>-admin.3scale.net" --name="my-unique-id" --service="123456789" --token="[redacted]"
+$ 3scale-config-gen --url="https://<organization>-admin.3scale.net" --name="my-unique-id" --service="123456789" --token="[redacted]"
```
I dont think you need to specify the --service and --token in config gen as it is now provided via annotations. @unleashed  Could you please confirm this?

unleashed replied:
The `--token` is needed (as it will always be part of the handler), but you are right that `service` can be omitted if it is provided via the annotations - there should already be some section explaining (and encouraging the user) to use the annotations, though. Perhaps we can add a note next to this example to mention that you can omit the `service` if you will use the annotations. Btw, @dfennessy as you also missed the important detail that this should be invoked from within the container image via `oc exec`, both notes could be grouped together next to this example.

dfennessy commented:
Thanks @pehala and @unleashed 

Updates made.

pehala commented:
LGTM!

dfennessy commented:
@neal-timpe 

This looks good from a 3scale perspective. Both SME and QE have approved. If anymore reviews are necessary, please feel free to share. 

Cheers.

dfennessy commented:
Hi @JStickler 

Neal suggested adding you as a peer reviewer. 

As you will see this has been approved from a 3scale SME and QE point of view.

I would appreciate approval from a Service Mesh team perspective. Cheers.

neal-timpe commented on modules/ossm-cr-example.adoc:
```diff
@@ -187,9 +187,9 @@ spec:
+    type: Mixer
     mixer: # only applies if policy.type: Mixer
-      enableChecks: false
+      enableChecks: true
```
@dfennessy Is this a change that is specific for 3scale? Or is this a change you meant to make in this example for the entire product?

JStickler replied:
+1 This is a service mesh example file, and I don't think we should be recommending using a deprecated component unless there is a need (that is, for 3scale).

dfennessy replied:
See point 3. I may have misread it: https://issues.redhat.com/browse/THREESCALE-6894?focusedCommentId=16253154&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16253154

dfennessy replied:
@neal-timpe  @JStickler 

Is the consensus is to change this to `false`?

JStickler replied:
Our comments are less about if the setting is correct and more that this setting only applies to 3scale and should be documented in the 3scale section of the docs and not in the general custom resource example.

dfennessy replied:
Ah, I see. It came up as part of the original review with the SME and has since been signed off by QE. My initial thought would be not to change anything for now.

@unleashed and/or @phala might have some comment on it, however they might not be available at the moment.

neal-timpe commented on modules/ossm-mixer-policy.adoc:
```diff
@@ -10,6 +10,7 @@ In previous versions of {ProductName}, Mixer’s policy enforcement was enabled
 .Prerequisites
 * Access to the {product-title} Command-line Interface (CLI) also known as `oc`.
 
+NOTE: `istio-system` is the default namespace the Service Mesh Control Plane (SMCP) installs to. You can change `istio-system` to any other name during the installation.
```
`istio-system` is not exactly the default. In the docs, we give istio system as the exampel as the control plane namespace. You could choose anything. But I'm not sure that you can change the name of it to anything else during the installation.

I think it would be more consistent with some changes I've recently made to say in step 2: 

>In this example `istio-system is the name of the `ServiceMeshControlPlane` namespace.

JStickler replied:
See my comment above about perhaps using angle brackets,   \<istio-system>  in the commands.

neal-timpe commented on modules/ossm-threescale-integrate-1x.adoc:
```diff
@@ -21,15 +20,17 @@ You can use these examples to configure requests to your services using the 3sca
 ====
 
-. Modify the handler configuration with your 3scale configuration.
+It is better to use a single handler per 3scale account by following the steps below:
```
Change this to:

>Use a single handler per 3scale account by following the steps below:

Same change for the 2.x version.

JStickler commented on modules/ossm-mixer-policy-1x.adoc:
```diff
@@ -10,6 +10,7 @@ In previous versions of {ProductName}, Mixer’s policy enforcement was enabled
 .Prerequisites
 * Access to the {product-title} Command-line Interface (CLI) also known as `oc`.
 
+NOTE: `istio-system` is the default namespace the Service Mesh Control Plane (SMCP) installs to. You can change `istio-system` to any other name during the installation.
```
There is no default namespace for the SMCP, customers create the namespace and then create the SMCP in that namespace.  A better note might be "The examples use `istio-system` as the control plane namespace.  Replace this value with the namespace where you deployed the SMCP."  Or even better, change the sample commands to show `istio-system` in angle brackets to indicate that it's a user configurable value.   \<istio-system>

JStickler commented on modules/ossm-threescale-integrate-1x.adoc:
```diff
@@ -21,15 +20,17 @@ You can use these examples to configure requests to your services using the 3sca
 [NOTE]
 ====
-Pay particular attention to the `kind: handler` resource. You must update this with your 3scale credentials and the service ID of the API you want to manage.
+Pay particular attention to the `kind: handler` resource. You must update this with your 3scale account credentials. You can optionally add a `service_id` to a handler, but this is kept for backwards compatibility only, since it would render the handler only useful for one service in your 3scale account. If you add `service_id` to a handler, enabling 3scale for other services requires you to crea ...
```
Last word should be `service_ids`  (no apostrophe)

JStickler commented on modules/ossm-threescale-integrate.adoc:
```diff
@@ -24,15 +23,17 @@ You can use these examples to configure requests to your services using the 3sca
 [NOTE]
 ====
-Pay particular attention to the `kind: handler` resource. You must update this with your 3scale credentials and the service ID of the API you want to manage.
+Pay particular attention to the `kind: handler` resource. You must update this with your 3scale account credentials. You can optionally add a `service_id` to a handler, but this is kept for backwards compatibility only, since it would render the handler only useful for one service in your 3scale account. If you add `service_id` to a handler, enabling 3scale for other services requires you to crea ...
```
Last word should be `service_ids` (no apostrophe)

JStickler reviewed (commented):
Couple of small updates, the most important being the change to the CR example.

JStickler commented:
Also, I see that this PR is against the 4.7 branch instead of master/main.  Do these changes not apply to 4.6, which is the EUS branch?  And is there some reason they should not be published to 4.8 branch once it goes live?

dfennessy commented:
> Also, I see that this PR is against the 4.7 branch instead of master/main. Do these changes not apply to 4.6, which is the EUS branch? And is there some reason they should not be published to 4.8 branch once it goes live?

@JStickler 

One of my initial questions before starting this was which branch I should create the PR against. My understanding was that I do it against 4.7...

dfennessy commented:
> > Also, I see that this PR is against the 4.7 branch instead of master/main. Do these changes not apply to 4.6, which is the EUS branch? And is there some reason they should not be published to 4.8 branch once it goes live?
> 
> @JStickler
> 
> One of my initial questions before starting this was which branch I should create the PR against. My understanding was that I do it against 4.7...

Can we cherry-pick to the required branch(es) in any case if necessary?

JStickler commented:
I think we can cherry pick, but I haven't tried it that way so I can't say for sure.  I haven't tried, but can you edit a PR to change it to make it against master?  It looks like it's possible from the Edit button next to the title of the PR.

dfennessy commented:
> I think we can cherry pick, but I haven't tried it that way so I can't say for sure. I haven't tried, but can you edit a PR to change it to make it against master? It looks like it's possible from the Edit button next to the title of the PR.

The edit is simple, however I get this, which makes me worry :sweat_smile: 

![image](https://user-images.githubusercontent.com/44813955/124000913-fcee5b80-d9cb-11eb-8890-a304dcd938fd.png)

JStickler commented:
Since you haven't touched your 3scale files in months I'd say you're probably OK, your master and 4.7 files are probably identical except for the changes you're making in this PR.  And OSSM team has cherry picked all our changes from master to the 4.7, so our files should be a close match too, if not identical between master and 4.7.  The old review comments part makes sense, because the diff might change between branches.  Not sure about how the commits might differ? I expect that would be a worry if you were a developer on a fast moving project.  But I think you're OK to make the change.

dfennessy commented:
> Since you haven't touched your 3scale files in months I'd say you're probably OK, your master and 4.7 files are probably identical except for the changes you're making in this PR. And OSSM team has cherry picked all our changes from master to the 4.7, so our files should be a close match too, if not identical between master and 4.7. The old review comments part makes sense, because the diff might change between branches. Not sure about how the commits might differ? I expect that would be a worry if you were a developer on a fast moving project. But I think you're OK to make the change.

Done...with conflicts :grimacing:

JStickler commented:
Oh good lord.  Try rebasing and see if that clears things up a bit.

dfennessy commented:
> Oh good lord. Try rebasing and see if that clears things up a bit.

I rebased and added my commit, however if doesn't looks like I've managed to clear up the conflicts. It appears to say I need write access.

I'll have to log off for the evening now and come back with fresh eyes in the morning...

If there is anything you think can be done to help fix this in the mean time, please share my way for tomorrow. Thanks for all your help Julie.
